src/io/signals.c

```diff
@@ -64,118 +64,233 @@ static const MVMAsyncTaskOps op_table = {
+    sig_vals[MVM_SIGXCPU]   = SIGXCPU;
 #endif
 #ifdef SIGXFSZ
-    case MVM_SIG_XFSZ:      signum = SIGXFSZ;   break;
+    sig_vals[MVM_SIGXFSZ]   = SIGXFSZ;
 #endif
 #ifdef SIGVTALRM
-    case MVM_SIG_VTALRM:    signum = SIGVTALRM; break;
+    sig_vals[MVM_SIGVTALRM] = SIGVTALRM;
 #endif
 #ifdef SIGPROF
-    case MVM_SIG_PROF:      signum = SIGPROF;   break;
+    sig_vals[MVM_SIGPROF]   = SIGPROF;
 #endif
 #ifdef SIGWINCH
-    case MVM_SIG_WINCH:     signum = SIGWINCH;  break;
+    sig_vals[MVM_SIGWINCH]  = SIGWINCH;
 #endif
 #ifdef SIGINFO
-    case MVM_SIG_INFO:      signum = SIGINFO;   break;
+    sig_vals[MVM_SIGINFO]   = SIGINFO;
 #endif
 #ifdef SIGUSR1
-    case MVM_SIG_USR1:      signum = SIGUSR1;   break;
+    sig_vals[MVM_SIGUSR1]   = SIGUSR1;
 #endif
 #ifdef SIGUSR2
-    case MVM_SIG_USR2:      signum = SIGUSR2;   break;
+    sig_vals[MVM_SIGUSR2]   = SIGUSR2;
 #endif
 #ifdef SIGTHR
-    case MVM_SIG_THR:       signum = SIGTHR;    break;
+    sig_vals[MVM_SIGTHR]    = SIGTHR;
 #endif
 #ifdef SIGSTKFLT
-    case MVM_SIG_STKFLT:    signum = SIGSTKFLT; break;
+    sig_vals[MVM_SIGSTKFLT] = SIGSTKFLT;
 #endif
 #ifdef SIGPWR
-    case MVM_SIG_PWR:       signum = SIGPWR;    break;
+    sig_vals[MVM_SIGPWR]    = SIGPWR;
 #endif
 #ifdef SIGBREAK
-    case MVM_SIG_BREAK:     signum = SIGBREAK;  break;
+    sig_vals[MVM_SIGBREAK]  = SIGBREAK;
 #endif
-    default:
+}
+
+static void populate_instance_valid_sigs(MVMThreadContext *tc, MVMint8 *sig_vals) {
+    MVMuint64 valid_sigs = 0;
+    MVMint8 i;
+
+    if ( tc->instance->valid_sigs ) return;
+
+    for (i = 0; i < NUM_SIG_WANTED; i++) {
+        if (sig_vals[i]) {
+            valid_sigs |=  1 << sig_vals[i] - 1;
+        }
+    }
+    tc->instance->valid_sigs = valid_sigs;
+}
+
+MVMObject * MVM_io_get_signals(MVMThreadContext *tc) {
+    MVMInstance  * const instance = tc->instance;
+    MVMHLLConfig *       hll      = MVM_hll_current(tc);
+    MVMObject    *       sig_hash;
+
+    MVMint8 sig_wanted_vals[NUM_SIG_WANTED];
+    populate_sig_values(sig_wanted_vals);
+
+    if (instance->sig_hash) {
+        return instance->sig_hash;
+    }
+
+    sig_hash = MVM_repr_alloc_init(tc, hll->slurpy_hash_type);
+    MVMROOT(tc, sig_hash, {
+        MVMint8 i;
+        for (i = 0; i < NUM_SIG_WANTED; i++) {
+            MVMString *key      = NULL;
+            MVMString *full_key = NULL;
+            MVMObject *val      = NULL;
+
+            MVMROOT3(tc, key, full_key, val, {
+                full_key = MVM_string_utf8_c8_decode(
+                    tc, instance->VMString, SIG_WANTED[i], strlen(SIG_WANTED[i])
+                );
+
+                key = MVM_string_substring(tc, full_key, 4, -1);
+                val = MVM_repr_box_int(tc, hll->int_box_type, sig_wanted_vals[i]);
+
+                MVM_repr_bind_key_o(tc, sig_hash, key, val);
+            });
+        }
+
+        instance->sig_hash = sig_hash;
+        populate_instance_valid_sigs(tc, sig_wanted_vals);
```

Comment:
Do we need to populate `valid_sigs` here?

There's a potential race condition if two threads are in this initialization at once; the first sets `instance->sig_hash`, then the second sees it is set in the condition above above and returns. The caller of the second then assumes `valid_sigs` is ready, but the first may not have finished setting it up yet. This might be avoided if we remove any expectation that this function sets up `valid_sigs`. Alternatively, I think swapping the order of the two lines above this review comment would also suffice.

Reply:
I added a check within the MVM_io_signal_handle function that checks that the valid_sigs value is populated prior to attempting to create the signal handler (at [line 288](https://github.com/MoarVM/MoarVM/pull/814/files#diff-a9d671f9f13e3d12535265ab82bde088R288)). Right now MVM_io_signal_handle is the only place where valid_sigs is being used.

The populate_instance_valid_sigs function is idempotent so it should not make a difference if multiple threads are running it simultaneously. That should be enough to guard against unpopulated valid_sigs, right?